packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -46,43 +56,62 @@ contract AmmPool is IBlockReceiver {
 
     event ExitPoolRequested(
         address  owner,
+        bool     toLayer2,
         uint     poolAmountIn,
         uint96[] minAmountsOut
     );
 
+    event QueueItemsProcessed(
+        uint numItems
+    );
+
+    uint public constant BASE = 10**18;
+    uint public constant INITIAL_SUPPLY = 100 * BASE;
+
     uint public constant MAX_AGE_REQUEST_UNTIL_POOL_SHUTDOWN = 7 days;
 
     IExchangeV3 public exchange;
     uint32      public accountID;
 
+    bytes32     public DOMAIN_SEPARATOR;
+
     uint        public shutdownTimestamp = 0;
 
+    enum PoolTransactionType
+    {
+        NOOP,
+        JOIN,
+        EXIT
+    }
+
     struct PoolJoin
     {
-        uint poolAmountOut;
+        address  owner;
+        bool     fromLayer2;
+        uint     poolAmountOut;
         uint96[] maxAmountsIn;
     }
 
     struct PoolExit
     {
-        uint poolAmountIn;
+        address  owner;
+        bool     toLayer2;
+        uint     poolAmountIn;
         uint96[] minAmountsOut;
     }
 
-    enum WorkType
+    struct PoolTransaction
     {
-        NONE,
-        JOIN,
-        EXIT
+        PoolTransactionType txType;
+        bytes               data;
+        bytes               signature;
```

Comment:
Seems in the current design, this signature field will always be empty. Shall we simply remove it?

Reply:
It most definitely is being used. Like with transactions on the exchange, it's cheaper to let users sign a request than have them do an onchain tx. Also more flexible for us, so seems win-win to use that by default. Only the onchain exit function is really needed for safety.

Only supporting joins from layer 2 + signature would makes things simpler because we don't have to worry about deposits/locking up funds.